### ISSUE TITLE
Source links for examples sometimes missing; all have disappeared from HTML output

### ISSUE THREAD

**Issue by lb42** [TEIC/TEI#1476]
It is required practice that as far as possible all examples in the Guidelines should not only be derived from real sources, but also cite those sources in the Bibliography.  This used to be checkable by means of a clickable link in the HTML display, associating the example with its entry in the bibliography, but that link seems to have disappeared. The link was derived from a @source attribute supplied on &lt;egXML> : in some cases that is missing (eg first example for &lt;argument>) ; in others it is present but not rendered.

**Comment by martindholmes** [comment]
Were existing `@source` attributes on `<egXML>` removed at some point, or have we simply got out of the habit of adding them?

**Comment by sydb** [comment]
595 of our 3236 `<egXML>` elements **do** have a `@source` attribute.

**Comment by hcayless** [comment]
Add a schematron warning recommending `@source` on egXML?

**Comment by laurentromary** [comment]
@hcayless : that would be specific to the guidelines. You would not want this for all  ODDs, would you?

**Comment by sydb** [comment]
Yes, the plan would be to put the constraint into TEIC/TEI/P5/p5odds.odd, which is used for validating the P5 source during the build process.

**Comment by sydb** [comment]
The Rump Council believes there may have been a mass loss of `@source`, so we should take a look at BIB entries, and see if there are a bunch which are not pointed at. SB will do so and report back.
We will hold off on the Schematron constriaint until we've checked on these. (Not really up to getting 1514 warnings.)

**Comment by sydb** [comment]
There are 18 `<bibl>`s in the bibliography that are **not** pointed to by an `<egXML>`; 6 of those are not pointed to by **anything**. Here is the list:

113139:29: bibl "DIC-CR" pointed to by ref/target
113335:31: bibl "FS-eg-01" pointed to by eg/corresp
113371:30: bibl "DIC-CED" pointed to by ref/target
113376:29: bibl "DIC-CP" pointed to by ref/target
113636:30: bibl "DIC-DNT" pointed to by ref/target
113663:31: bibl "CH-eg-01" pointed to by NOTHING
113667:31: bibl "DIC-OALD" pointed to by ref/target
113837:32: bibl "STGA-eg-9" pointed to by NOTHING
113850:32: bibl "DIC-LDOCE" pointed to by ref/target
113957:31: bibl "DIC-NPEG" pointed to by ref/target
114015:30: bibl "DIC-PLC" pointed to by ref/target
114060:29: bibl "DIC-PR" pointed to by ref/target
114081:29: bibl "SVG-11" pointed to by ref/target
114234:36: bibl "MENTIOND-eg-1" pointed to by NOTHING
114276:29: bibl "DIC-W7" pointed to by ref/target
114435:35: bibl "fr-ex-Fleurs" pointed to by NOTHING
114577:27: bibl "rab3" pointed to by NOTHING
114672:36: bibl "biblzh-tw_n47" pointed to by NOTHING

**Comment by lb42** [comment]
So the situation is not as bad as we feared in that we probably haven't lost links; it is however bad news that such a large number of examples don't have citations. Would be worth a systematic look.

**Comment by sydb** [comment]
Agreed. I think a quick historical check is in order, but unless that shows a mass removal, back to plan "A".

**Comment by lb42** [comment]
The Xpath //*:egXML[not(@source)] applied to current p5subset.xml returns 1580 results. Scrolling through, I note that a large number of them seem to be appropriately non-sourced, for example because they are made up examples demonstrating how to do something in the header.

**Comment by sydb** [comment]
Well, you probably want to search against p5.xml, not p5subset.xml, as there are lots of examples that are not in the tagdocs. But either way, yes, there are a lot that should have `source="#NONE"`.

**Comment by sydb** [comment]
LB to take first crack at assigning sources to those of the 2641 that actually have sources.

**Comment by lb42** [comment]
There aren't that many looking just at ones with non-whitespace text() children.

**Comment by lb42** [comment]
Attached stylesheet selects all unsourced egXMLs inside English language exemplums in modules other than the header : running it against p5subset produces the attached (fairly manageanble) set of things to look at. 
[Curses: for some reason git wont let me attach a zip file, or anything else to this ticket]

**Comment by jamescummings** [comment]
@lb42 Umm...again why are you running it against p5subset? As Syd pointed out above https://github.com/TEIC/TEI/issues/1476#issuecomment-249848402 you need to do it on P5 as a whole right?  There are egXML in the chapter prose which does not exist in p5subset.

**Comment by lb42** [comment]
Just dividing and conquering... most of the missing ones seem to be in specs.

**Comment by lb42** [comment]
Tracked down and patched up quite a few, by the simple expedient of scanning the source in oXygen. I started, but didn't finish, linking ones that are known to have no source because they are made up to the BIB entry intended for that purpose, but desisted pending a decision as to whether or not this should be specially handled in the HTML formatting.

**Comment by emylonas** [comment]
There are 467 egXML chunks that have no text content and have no source attribute - they are made up completely of subelements and attributes. They all appear to be made up, although it's possible that some derive from the encoding or other declarations of some project or other. 

We can automatically mark them all as having `source="#NONE"`
We can, alternatively, mark them all as having  `source="#UND"` (undetermined) and create a new bib entry to point to if necessary.
We can, finally, leave these without an `@source` attribute, and move on to the ones that do have content. 

It doesn't seem worth it to examine each independently and to look for sources for these egXMLs.

**Comment by emylonas** [comment]
Unless we determine otherwise in Council phone call 10/27 we will implement solution number 2

**Comment by sydb** [comment]
Done at f6b4a55.

**Comment by sydb** [comment]
@martindholmes and I suspect this may have been what broke the TEI-Documentation-dev build. Hope to look into further later today or tomorrow.

**Comment by hcayless** [comment]
Hope I've fixed all the build-breaking backlinks with some tweaks to the Stylesheets in TEIC/Stylesheets@65505f7

**Comment by hcayless** [comment]
Can we close this issue?

**Comment by sydb** [comment]
Well, by my count there 2069 remaining `<teix:egXML>` elements that do **not** have a `@source`. So I would say no, we shouldn’t close the ticket. What we _do_ with those 2069 examples is another story.

**Comment by emylonas** [comment]
This is a tedious task. Also not a high priority task. Although no longer on Council, I'm willing to hang on to it.
The best way to attack this problem is 
1. to go through all the examples and identify the low hanging fruit.

**Comment by lb42** [comment]
As I started this, I am happy to contribute some effort Eli!

**Comment by lb42** [comment]
Some egXMLs have no value for @source; some have a value #UNDEF; some have a value #NONE. I suppose the purpose of the last is to distinguish known unknowns (i.e. we tried, but we know that there is no source for this because we made it up), but why do we have the first two? Is it to distinguish cases where  we have tried to find a source and failed from ones where we have not yet bothered to look?  In any case, I am pretty sure this complex three way distinction is inconsistently applied and not very useful.   Should the stylesheets not simply be modified so as NOT to generate the "bibliography" link and its associated but information-free pop up in each of these cases?

**Comment by sydb** [comment]
You are absolutely correct (except that #UNDEF is actually just #UND). No `@source` is simply an error; #UND is undetermined — we have put some effort into finding a source, but have (as of yet) been unsuccessful; #NONE is for examples that were simply dreamt up by somebody; in addition there are perhaps half a dozen cases of #SELF, meaning the example is copied from the _Guidelines_ themselves.

While I suspect you can find some cases of inconsistency, since the vast majority of this work was done by only two people (myself and @emylonas), I hope it’s not **too** bad.

I daresay these values are very important to those of us trying to maintain the bibliographical citations for examples, and pretty close to useless for a typical reader of them. So I would be pretty happy with a Stylesheets change that ignored at least  #NONE and #UND, or perhaps better, compressed all 500+ of those little arrows into one.

**Comment by lb42** [comment]
Sorry I didn't mean to sound unappreciative of the work already done, but I am just unpersuaded that the distinction between #UND and not(@source) is actually useful. All the examples are "dreamt up by somebody"; in some cases we can identify the source, in others either we can't and we know we can't  (so #NONE) or we haven't yet made that determination (so either UND or silence). As there are nearly four times as many in the latter category -- no @source--  as the former -- @source=UND --  the usual rule of simplification would be to abolish #UND. Why do you think the absence of @source is an error? 

FWIW, here are the current counts:

#UND  : 500
#SELF : 4 (one of which is not actually a quote from the GL)
#NONE : 225
no @source: 1935

**Comment by sydb** [comment]
Well, if you want to go have a look at those 500 #UND cases, have at it! (And I will not be at all surprised at all if you can figure out a few that @emylonas or I could not). 

And yes, they’re all dreamt up by someone. But some examples are lifted by the TEI lock, stock, and encoding from an existing project; some are modified by the TEI from an existing project; for some the TEI creates an encoding for an existing work; and for some the TEI invents both the work and the encoding for it. It is only that last category that I think of as #NONE.

The absence of `@source` is an error only in that Council decided every `<egXML>` in the _Guidelines_ should have an `@source`. The schema (P5/p5odds.odd) does not (yet?) actually require it.

**Comment by sydb** [comment]
And which of the 4 #SELF is not actually a self-quotation?

**Comment by lb42** [comment]
The question I am raising is what's the difference between source="" and source="UND" and my claim is that there isn't any. 

In my view,  examples taken from an existing project, whether modified or not,  should cite the project (don't know if there are any such). Examples which represent the way some existing work is or might be encoded should cite the existing work. Examples for which there is no source because they are purely hypothetical should have source="".

**Comment by sydb** [comment]
I presume you mean the difference between `not(@source)` and source="#UND", right? My claim is that the latter means someone (probably @emylonas, but perhaps me and possibly someone else) has tried to find a source and failed.

>  examples taken from an existing project, whether modified or not, should cite the project

Agreed, although I think typically we’ve been citing the project’s version of the work, not the whole project. And one caveat: if the project is the TEI-C and the work is the _Guidelines_, I’ve been using #SELF.

> Examples which represent the way some existing work ... might be encoded should cite the existing work.

Agreed; and that’s what I believe we have been doing.

> Examples for which there is no source because they are purely hypothetical should have source="".

No, they should have source="#NONE", IMHO, to make this clear.

**Comment by emylonas** [comment]
Agree with Syd. Anything added from now on should have a source that is either cited or `#NONE`, which means it's completely made up. We could come up with a citation for that, too, such as `#FABRICATED`.  The `not(@source)` examples are ones that have not been checked. They may have sources or may not. `#UND` means that there is possibly a source, but we don't know it. So no source is not the same as undetermined source, at the moment, but it may be once this process is complete. One common problem are examples that cite markup from that was taken from someone's actual project or corpus, but we don't know which.

**Comment by lb42** [comment]
I would use source=NONE for all invented examples. What advantage/rationale is there in distinguishing "fabricated" from "none"? Clearly if there's no source the example must have been fabricated!

**Comment by lb42** [comment]
@sydb my mistake -- on closer inspection, all four cases of source=SELF are indeed quotations from the Guidelines source. (A pedant might argue that *all* the examples are quotations from the Guidelines but, as you know, I am not a pedant)

**Comment by ebeshero** [comment]
This ancient ticket requires some investigation to see what (if any) issues remain to be solved.

**Comment by sydb** [comment]
Current counts (in p5.xml, 4.10.0a, last updated 2025-05-16, rev a449704c0):

- total `<egXML>` elements = 3494
  - with `@source` = 1474
     - source of #UND = 505
     - source of #NONE = 240
     - source of #SELF = 6
   - **without** `@source` = 2020
